### PR TITLE
Columns: Update columns to allow full height stretch when vertically aligned

### LINKED ISSUE #32907: Columns block: Background color does not stretch full-height with vertical alignment set
## Description
Background color does not stretch full column height with a vertical alignment set.

## Step-by-step reproduction instructions
1. Open the editor and insert the Columns block. Pick the two-column 50-50 layout.
2. In the first column insert an image. In the second column insert a heading & paragraph.
3. Set the background color AND vertical alignment of the second column.
4. Note the background does not stretch full column height, but only behind the column content.
5. Reset the column vertical alignment. Note the background now stretches the full height of the column.

## Expected behaviour
I expected background color to cover full column height, regardless of vertical alignment of the content inside.

## Screenshots or screen recording (optional)
Preview - no alignment on the right column
image

Preview - right column set to align middle
image

Editor - right column set to align middle
image

## WordPress information
- WordPress version: 5.7.2
- Gutenberg version: 10.8.2
- Are all plugins except Gutenberg deactivated? Yes
- Theme: Blank Canvas & TT1

## Device information
- Device: Desktop, macOS 11.4 / Chrome 91

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/32907

## Description

- Allows column blocks to stretch to the full height when aligned vertically so background colors appear as expected.

#### Notes

The ability to set a background color on individual columns was recently added. Prior to then columns simply set `align-self` styles to achieve the vertical alignment.

This PR changes columns to be `display: flex;` and vertically align their children within them in line with the column's vertical alignment selection. This is working with multiple inner blocks or nested columns. 

## How has this been tested?
Manually.

1. Checkout this PR, create a new post and add a columns block, choosing the 50/50 split
2. Within the first column insert a tall image or block of content.
3. Select the second column, set a background color and add some short content
4. Try out the different vertical alignment options for the second column
5. Save post and confirm display is correct on the frontend
6. Switch themes and reconfirm display is correct

## Screenshots 

| Before | After |
|---|---|
| Screen Shot 2021-06-23 at 5 45 35 pm | vertical-align-column-fix |

## Types of changes
Bug fix / enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by paaljoachim** [comment]
Associated Gutenberg support issue: https://wordpress.org/support/topic/column-equal-height

**Comment by andrewserong** [comment]
Thanks for working on this one @aaronrobertshaw, I like this change, and it feels like a better approach for the alignment (we want to align the child items of the column rather than the column itself). Unfortunately it looks like this introduces an issue for horizontal alignment with blocks like the image block.

Before this change, with the image left aligned, the inserter or block to the right of the image will move around the image:

https://user-images.githubusercontent.com/14988353/124233424-2b983d80-db56-11eb-8932-1c83acaa74fb.mp4

After this change, the inserter doesn't move relative to the size of the left aligned image, and when I insert a paragraph block, it sits behind the left aligned image:

https://user-images.githubusercontent.com/14988353/124233511-423e9480-db56-11eb-939e-235d0ee79a5a.mp4

I'm guessing that the float-based alignment doesn't play nicely with being a flex item?

Other than that, the change works nicely for retaining the height of the column and the column background while aligning the items within it. I can't quite think of a simple workaround without adding an additional container?

**Comment by aaronrobertshaw** [comment]
Appreciate the detailed review as always @andrewserong 👍 

> Unfortunately it looks like this introduces an issue for horizontal alignment with blocks like the image block.
> 
> Before this change, with the image left aligned, the inserter or block to the right of the image will move around the image:

I actually see different behaviour when I run trunk. The alignment of an image block as an immediate child of the column doesn't work there either. 

This is what I experience with the current layout/styles on `trunk`:

ColumnAlignmentIssuesTrunk

> I'm guessing that the float-based alignment doesn't play nicely with being a flex item?

That's my guess as well as to the root cause. The issue on trunk might have similar origins just further up the tree.

I've pushed a comment that forces an auto margin left on an immediate child of the column with the `data-align="right"` attribute. This pushes it to the right but will not allow wrapping of other content. The wrapping content can still be achieved though by using a group block around the content the user wishes to float and wrap. Essentially, that is the additional container you mentioned.

After applying the new style to force right alignment I get the following behaviour:

ColumnAlignmentIssues4

I suspect we might also have some theme specific issues with this, so it could use further testing on that front. Which theme were you using in your videos? I couldn't replicate the issue with the image's resizable container overlapping the quick inserter.

**Comment by aaronrobertshaw** [comment]
I've found that I can replicate the issue with the inserter being overlapped when using the Maywood theme. I'll test further with that.

**Comment by aaronrobertshaw** [comment]
It appears that the overlapping issue comes from the `.wp-block[data-align]` element getting a height of `0` via classic.css in the theme. Switching that to `height: auto` makes it behave more like a theme such as TwentyTwentyOne Blocks.

Looking into this has highlighted the change to flex display also prevents the inserter DOM element getting a width resulting in the inserter icon display off to the left of where it should be. This effects all themes for me.

I've tried switching a few different approaches using flex or grid without a clear winner. A few additional tweaks have been pushed, along with rebasing this one.

Let me know if you have any ideas as to an alternate approach, I'm running out of ideas.

**Comment by andrewserong** [comment]
Thanks for digging into this one @aaronrobertshaw (and figuring out that I was using the Maywood theme!). It _is_ a tricky one.

> The wrapping content can still be achieved though by using a group block around the content the user wishes to float and wrap. Essentially, that is the additional container you mentioned.

As a workaround, that actually sounds quite reasonable to me. Since this PR could introduce a regression for existing layouts that folks might be using with wrapped content, I'm wondering if it'd be worth moving the `display: flex` to only apply when one of the vertical alignment options is selected? E.g. move it to the `&.is-vertically-aligned-top` rules, etc? There'd be a bit of duplication adding it to the separate rules, but it might help guard against existing Columns out in the wild being adversely affected?

On my test site it's working nicely using an additional group block for the situation where I might need the text to wrap around the aligned image element: (The unusually large vertical margin is from the Maywood theme's styling):

| Left alignment within column | Left alignment within group block within column |
| --- | --- |
| image | image |

> I've tried switching a few different approaches using flex or grid without a clear winner.

I think this feature is a matter of striking a balance between trade-offs, unfortunately! Being able to have the column background color reach the full height of the columns block unlocks a number of good options for pattern designs (e.g. vertically centered content in a row of columns with striking background colors) that we mightn't be able to achieve otherwise. And with this change, it's still possible to do the floated alignment with wrapped paragraphs if the user adds an additional container block. So, I think if we can see about only setting the column to flex when a vertical alignment is applied, hopefully that'll be a decent enough trade-off.

We'll then want to test this across a number of themes / block combinations to make sure we're happy with the trade-offs?

One other tiny note, it looks like this needs a rebase now after https://github.com/WordPress/gutenberg/pull/31816 landed (apologies!)

**Comment by aaronrobertshaw** [comment]
> Since this PR could introduce a regression for existing layouts that folks might be using with wrapped content, I'm wondering if it'd be worth moving the display: flex to only apply when one of the vertical alignment options is selected?

This is actually a thought I had while revisiting this as well. I missed adding the style back in after continuing to experiment, looking for alternative solutions. Thanks for catching this!

> I think this feature is a matter of striking a balance between trade-offs, unfortunately!

Agreed. I believe this is good option for the moment as it does, as you say, unlock a lot of design possibilities while still allowing the wrap around effect albeit with the use of the group block.

> One other tiny note, it looks like this needs a rebase now after #31816 landed (apologies!)

I rebased this prior to adding back in the style to only apply flex layout when column is aligned vertically.

**Review by andrewserong — APPROVED** [review]
This is testing nicely for me @aaronrobertshaw, and in a multi-column layout where a user might want an active vertical alignment setting on one or more columns, while keeping one of the columns without that alignment setting, the floated left/right alignment is working correctly. For example, this kind of layout works (though the Columns blocks' mobile breakpoints don't play that well with it):

image

I think the current implementation in this PR strikes a good balance of trade-offs, enabling the desired styling, while minimising impact for left/right alignment, so this LGTM, and I tested with half a dozen themes across Safari, FF, Chrome, and Edge. But, it might be worth getting another pair of eyes from a designer, too 🙂

**Comment by aaronrobertshaw** [comment]
Appreciate the thorough testing across themes and browsers @andrewserong. 🙇  

It is certainly appreciated given the potential impact of such changes. To that end, I agree, it would be handy to get some extra eyes on this.

@jasmussen or @paaljoachim if you have some spare time, would you be able to take a look at this and let me know if there is anything that I've missed. Thanks in advance 👍

**Comment by retrofox** [comment]
Hi, there. Have you considered adding a Full height aligment option to the toolbar Something that [we did for the cover block ](https://github.com/WordPress/gutenberg/pull/26615).

**Comment by aaronrobertshaw** [comment]
Thanks for the suggestion @retrofox 👍 

I did consider a similar approach early on. It felt like it was adding extra complexity to the controls and toolbar when the user is already more than likely requesting full height behaviour. The issues @andrewserong highlighted in regards to wrapping inner aligned/floated elements would remain either way.

My understanding of our general approach to these sorts of issues is to err on the side of providing sensible defaults and minimising the need for additional controls. It feels as though full height columns when a background is set is a fairly sensible or expected behaviour that also avoids the need for an extra control.

I'm more than happy to get further opinions on this and update as needed or include the extra control/classes etc.

**Comment by jasmussen** [comment]
Coming back to this after some AFK, thanks for the ping! Thanks for working on this, it's a fix we want. My immediate and only question is this part of the code:

```
	// Only apply display flex for vertically aligned columns.
	&[class*="is-vertically-aligned-"] {
		display: flex;
		flex-direction: column;
	}
```

It sounds to me like floats are the primary challenge that makes this be a conditional, correct? Specifically thinking of this comment:

> The wrapping content can still be achieved though by using a group block around the content the user wishes to float and wrap. Essentially, that is the additional container you mentioned.

Floats have been challenging in flex containers in the past, and a recent `layout` related refactor fully embraced and suggested the "group first" approach to floats (something which we can potentially enhance further with patterns and pattern transforms). 

Mostly I'm concerned about code complexity and a proliferation of edgecases by changing the display mid-stream like this. Do these concerns make sense or am I missing a beat? Thanks @andrewserong and @aaronrobertshaw

**Comment by jasmussen** [comment]
Here's an adjacent conversation about floats: https://github.com/WordPress/gutenberg/issues/33332#issuecomment-877251979:

> > Could a possible solution involve making the LayoutStyle component less opinionated about which blocks it applies styles to?

> That doesn't sound like a great solution because alignments are dependent on their parent, for instance in a "columns" block (flex container), floats are not allowed.

**Comment by andrewserong** [comment]
> It sounds to me like floats are the primary challenge that makes this be a conditional, correct?

Yes, I believe so.

> Mostly I'm concerned about code complexity and a proliferation of edgecases by changing the display mid-stream like this. Do these concerns make sense or am I missing a beat?

Those concerns definitely make sense to me! I know what you mean, each time we add edge cases, it makes the CSS harder to read and means we've got additional cases to test whenever we make changes. So good to avoid if possible.

It's a tricky one with the Column block, because I do like the idea of ditching floats altogether and just letting flex work the way it needs to, but we also want to avoid causing regressions for columns out in the wild where folks might have used left or right aligned images as direct children of the column. So, I think the trade-off here is how much backwards compatibility we want to maintain (and whether we're happy with the potential impact of switching the column to flex).

@jasmussen would your preference be for us to keep flex and ditch the rules to support floats, or do you have another approach in mind?

Damian's suggestion of a full height alignment option is also a compelling idea, but could also wind up adding extra complexity, too? 🤔

**Comment by jasmussen** [comment]
It's a doozy, I see the problem. Each _column_ is `display: block;` which means if it's aligned using "align-self", then it no longer spans the height. Whereas in this branch, it uses justify-content, to align the insides instead of the block itself, thus requiring flex. 

The curious thing here is, in my testing (block themes only), floats don't work at all inside columns even now. Which we either embrace and disable the float button, or fix with column specific float overrides such as this PR adds. 

But if we do that, add float overrides, couldn't we just embrace flex for column's (singular column blocks) entirely, and remove the conditional?

**Comment by andrewserong** [comment]
> The curious thing here is, in my testing (block themes only), floats don't work at all inside columns even now. Which we either embrace and disable the float button, or fix with column specific float overrides such as this PR adds.

Interesting! Re-testing on trunk now, I can't reproduce floats working with paragraph text wrapping around floated image blocks within a column block, either. So if that's the current behaviour, then yes, maybe it'd be better to embrace flex entirely. My only reservation is making sure that we don't accidentally introduce regressions for columns that are out in the wild.

But I agree, it'd be nice to embrace flex entirely if we can. We've been focused on some other PRs this week, but I'm happy to do a bit more testing on this next week to see what's feasible!

**Comment by aaronrobertshaw** [comment]
> Interesting! Re-testing on trunk now, I can't reproduce floats working with paragraph text wrapping around floated image blocks within a column block, either.

I believe your earlier testing was with the Maywood theme @andrewserong.

On trunk, quickly flicking the theme over to Maywood, you can see an image floats still within a column allowing the following paragraph block to wrap. Other themes out of my limited library that allowed the float wrapping were; Ambitious, Hever, Seedlet, TwentyTwentyOne (non blocks), TwentyTwenty, TwentyNineteen and so on.

Screen Shot 2021-08-06 at 5 44 16 pm

I like the consistency of embracing flex and dropping support for floated elements as immediate children of the column. There should still be the work around of enclosing whatever you want to wrap within a group block.

That still leaves us with the problem of backwards compatibility for all the themes out there that use floats for left/right alignment. The current CSS rule limiting the display of flex to only vertically aligned columns does reduce the potential breakage for columns and themes already out in the wild. As much as I'd like to drop it, I don't think we can.

**Comment by andrewserong** [comment]
Thanks for the reminder, Aaron, yes I was testing with Maywood.

> The current CSS rule limiting the display of flex to only vertically aligned columns does reduce the potential breakage for columns and themes already out in the wild. As much as I'd like to drop it, I don't think we can.

Agreed, it does feel safer to preserve that.

**Comment by aaronrobertshaw** [comment]
> That doesn't sound like a great solution because alignments are dependent on their parent, for instance in a "columns" block (flex container), floats are not allowed.

The previous testing of various themes would indicate to me, that comments elsewhere stating floats are not allowed in columns isn't exactly accurate or at least is ambiguous. 

Whether columns should be exclusively flex containers and not support float alignments of inner blocks, is perhaps a separate question. I think of this PR as correcting or bringing the behaviour more into line with existing use and expectations, rather than drawing a line in the sand over how columns should be laid out.

Can this land as is, given it does improve the columns block? Would it prevent a more detailed discussion of converting columns to flex only in a separate issue/PR?

**Comment by jasmussen** [comment]
> Can this land as is, given it does improve the columns block? Would it prevent a more detailed discussion of converting columns to flex only in a separate issue/PR?

Since it would be easy to follow up on, nothing beats trying it out in practice. 👍 👍

**Comment by apeatling** [comment]
@aaronrobertshaw Looks like this one could merge, are you waiting on any more feedback?

**Comment by aaronrobertshaw** [comment]
Thanks for checking in @apeatling. I was waiting to see if the work around flex via the layout support might have been adopted for the Columns block. 

At this point, if there are no objections to the current approach, I'll rebase it, give it another test then can merge it.

**Comment by aaronrobertshaw** [comment]
After some further testing on this I found;

- the original issue remains a problem on trunk
- the approach taken in this PR still fixes the column height after a rebase
- a new issue where a column's immediate child had narrow content it would no longer fill the column width

Screen Shot 2021-09-22 at 11 55 56 am

To address the new issue I've added a `:where` CSS rule to assign a width of `100%` to immediate child `.wp-block` elements within a vertically aligned column. This solves the problem for me while remaining easy enough to override when needed.

#### Before (Trunk)

https://user-images.githubusercontent.com/60436221/134272517-f3a9450c-b179-4778-8341-831aba9b3d3c.mp4

#### After (This PR)

https://user-images.githubusercontent.com/60436221/134272549-f7bdd24d-0b4e-4a17-9a39-29337d7ae6c7.mp4

**Comment by andrewserong** [comment]
Thanks for following up on this one @aaronrobertshaw! I just took it for a spin, and noticed that with this PR applied, adjusting the vertical alignment also appears to affect the spacing between blocks. Here's a before / after:

| Before (trunk) | After (this PR) |
| --- | --- |
| Kapture 2021-09-22 at 12 35 12 | Kapture 2021-09-22 at 12 39 19 |

I'm testing with ~TT1-Blocks theme active~, but let me know if you need any extra info! Edit: I had Twenty Twenty-One (not TT1-Blocks) active.

**Comment by aaronrobertshaw** [comment]
Appreciate the quick turnaround in testing @andrewserong 👍 

> adjusting the vertical alignment also appears to affect the spacing between blocks

I don't believe this one will be theme-specific.

We are having to rely on setting a flex layout for the column. As far as I'm aware, flex layouts don't support margin collapsing and it is the lack of margin collapsing that results in the increased spacing between blocks. In other words, all the flex items' margins are actually honored.

I'm not having much luck working around this. So open to fresh ideas if you have any?

**Comment by andrewserong** [comment]
> I'm not having much luck working around this. So open to fresh ideas if you have any?

Thanks for the explanation surrounding margin collapse, that makes sense why it's behaving differently! I just did a bit of hacking around, but unfortunately I haven't been able to come up with a workaround, either. It'd be a pity to mess with the vertical rhythm of an existing theme, but I suppose at least the behaviour is guarded behind having a vertical alignment selected? I think the thing to explore next is whether or not the trade-off is appealing to designers or folks with a bit more experience working with theme development.

Some not particularly useful dead-end ideas:

* See if we could do padding-top or padding-bottom instead (doesn't really work, because we can't know what value to set it to without some JS).
* Adding an additional container to the column (kinda works, but goes against the approach for blocks to keep the markup lean).
* Add a group block within each column block (to an inner blocks template, perhaps?) — works manually, but like adding another container, it adds to the overall nesting which isn't great.

Overall, I like the idea of using flex for the column (and it _feels_ like the right way to go about it), but it'll be interesting to see how much themes are relying on the margin collapse and whether or not it's a deal breaker 🤔

**Comment by ramonjd** [comment]
Sep-22-2021 15-15-51

Probably highly dodgy, but we could play with `row-gap` for the above effect. 

Here's what I did to get the above effect:

```
	/**
	* Individual Column Alignment
	*/
	&.is-vertically-aligned-top {
		justify-content: flex-start;
		row-gap: var(--global--spacing-vertical) !important;
	}

	&.is-vertically-aligned-top > * {
		margin-top: 0 !important;
		margin-bottom: 0 !important;
	}
```

P.S. I used `!important` only because I was `!impatient` in my testing 😄

**Comment by aaronrobertshaw** [comment]
> Probably highly dodgy, but we could play with row-gap for the above effect.

An interesting idea. My concern would be how that would play with blocks that the user has set manual margins on e.g. via blocks supports, or maybe a theme author by theme.json where the value doesn't match the global.

**Comment by ramonjd** [comment]
I was testing this again last night and couldn't find a way around the lack of margin-collapsing for flexbox. But it made me think: what if we gave column blocks a `display: flex` with (potentially) a vertically-aligned value of top by default?

**Comment by aaronrobertshaw** [comment]
> I was testing this again last night and couldn't find a way around the lack of margin-collapsing for flexbox

I haven't found the solution yet either. 

Once we have increased adoption of block gap or margin support, the extra spacing due to the lack of margin-collapse for flexbox could be configured away. I think this is an acceptable price to pay for users wanting to be able to have the full height background in columns. 

Even as it is now, if that extra spacing is a problem, you could wrap contents in a group block and regain that margin-collapse. Obviously, it's difficult to expect users to know that's an option. It is for an edge case and the desire to have full height backgrounds appears to be of more concern.

>  But it made me think: what if we gave column blocks a display: flex with (potentially) a vertically-aligned value of top by default?

Doing this would potentially introduce a regression on blocks already out in the wild. In particular, those columns that have floated/aligned inner content. https://github.com/WordPress/gutenberg/pull/32909#issuecomment-876841766

**Comment by porg** [comment]
## No workaround for this design need with native blocks

- To use a group in row layout with vertical alignment "Stretch to fill" achieves the effect of same height blocks.
- But then there's no responsive wrapping possible (that they get full width on mobile).

## PR fix attempt already 3 midsummers old now…

- Would appreciate that function still.
- Has the root cause maybe a fix meanwhile?

**Comment by aaronrobertshaw** [comment]
Closing this given the [recommendation](https://github.com/WordPress/gutenberg/issues/32907#issuecomment-5312465168) on the original issue. 

**TL;DR** use an Row > (Image + Stack) combination of blocks to achieve the same full height background color.
